ambry-router/src/main/java/com.github.ambry.router/DeleteOperation.java

```diff
@@ -139,19 +147,32 @@ private DeleteRequest createDeleteRequest() {
    * that different {@link ServerErrorCode} are received from different replicas. These error codes
    * are eventually resolved to a single {@link RouterErrorCode}.
    * @param responseInfo The response to be handled.
+   * @param shouldOnlyRecordMetrics {@code true} indicates that only update metrics for the corresponding
+   *                                            metrics, and do not handle the response.
    */
-  void handleResponse(ResponseInfo responseInfo) {
+  void handleResponse(ResponseInfo responseInfo, boolean shouldOnlyRecordMetrics) {
     DeleteRequest deleteRequest = (DeleteRequest) responseInfo.getRequest();
-    InflightRequestInfo inflightRequestInfo = inflightRequestInfos.remove(deleteRequest.getCorrelationId());
-    // inflightRequestInfo can be null if this request was timed out before this response is received.
-    if (inflightRequestInfo == null) {
+    DeleteRequestInfo deleteRequestInfo = deleteRequestInfos.remove(deleteRequest.getCorrelationId());
+    // deleteRequestInfo can be null if this request was timed out before this response is received. No
+    // metric is updated here, as corresponding metrics have been updated when the request was timed out.
+    if (deleteRequestInfo == null) {
+      return;
+    }
+    routerMetrics.routerRequestLatencyMs.update(time.milliseconds() - deleteRequestInfo.getStartTimeMs());
+    ReplicaId replica = deleteRequestInfo.replica;
+    DataNodeId dataNodeId = replica.getDataNodeId();
+    NonBlockingRouterMetrics.NodeLevelMetrics dataNodeBasedMetrics =
+        routerMetrics.getDataNodeBasedMetrics(dataNodeId);
+    if (shouldOnlyRecordMetrics) { // Indicate the corresponding operation has been already completed.
```

Comment:
This seems awkward. Could you try to refactor this code so that metric updates do not come in the way of core logic? I would advise against being influenced by coordinator code too much.